core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java

```diff
@@ -653,40 +652,33 @@ public void closeUfsBlock(long sessionId, long blockId)
   }
 
   @Override
-  public BlockReader createBlockReader(BlockReadRequest request) throws
-      BlockDoesNotExistException, IOException {
-    long sessionId = request.getSessionId();
-    long blockId = request.getId();
+  public BlockReader createBlockReader(long sessionId, long blockId, long offset,
+      boolean positionShort, Protocol.OpenUfsBlockOptions options) throws IOException
+  {
     RetryPolicy retryPolicy = new TimeoutRetry(UFS_BLOCK_OPEN_TIMEOUT_MS, Constants.SECOND_MS);
-    while (retryPolicy.attempt()) {
-      try {
-        BlockReader reader = createLocalBlockReader(sessionId, blockId, request.getStart());
-        Metrics.WORKER_ACTIVE_CLIENTS.inc();
-        return reader;
-      } catch (BlockDoesNotExistException e) {
-        boolean checkUfs =
-            request.isPersisted() || (request.getOpenUfsBlockOptions() != null && request
-                .getOpenUfsBlockOptions().hasBlockInUfsTier() && request.getOpenUfsBlockOptions()
-                .getBlockInUfsTier());
-        if (!checkUfs) {
-          throw e;
+    if (hasBlockMeta(blockId)) {
+      while (retryPolicy.attempt()) {
+        try {
+          BlockReader reader = createLocalBlockReader(sessionId, blockId, offset);
+          Metrics.WORKER_ACTIVE_CLIENTS.inc();
+          return reader;
+        }
+        catch (BlockDoesNotExistException e) {
+          LOG.error("Block should exist in local store but not: ", e);
+          throw new InternalException("Block should exist in local store but not", e);
```

Comment:
It could have a concurrency problem since we doesn't lock block after hasBlockMeta before createLocalBlockReader. So it could happen that we delete block in between. We should add some comment about this

Reply:
This logic is different from before. I was debating whether this is a good change or not. With the previous logic, we will try UFS is `chckUfs` is true. But I kinda feel that the behavior is just hiding issues so I changed it to always throw here. I don't know why the original logic is written that way. If it is to handle the condition you mentioned, then I think the right way to handle it is to lock the block. If we don't know why we would check ufs (which is me when I wrote this), then I'd rather have a loud failure and get a better understanding of how likely this would happen, than quietly trying UFS.

Reply:
Reverted to the old logic.